walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,174 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
 
 public final class FindMeetingQuery {
+  private List<TimeRange> attendeesEventTimes;
+  private Map<String, List<TimeRange>> optionalAttendeesEventTimes;
+  private List<List<String>> optionalAttendeesCombs;
+
+  /** 
+   * Finds a list of times when the requested event can happen, 
+   * given the list of current events and the request information. 
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> attendees = new ArrayList<>(request.getAttendees());
+    Collection<String> optionalAttendees = new ArrayList<>(request.getOptionalAttendees());
+    // On the webapp, if form is left empty, an empty string is obtained. Remove them here.
+    attendees.removeAll(Collections.singleton(""));
+    optionalAttendees.removeAll(Collections.singleton(""));
+
+    long duration = request.getDuration();
+
+    // Populate {@code attendeesEventTimes} and {@code optionalAttendeesEventTimes}.
+    getAttendeesEventTimes(events, attendees, optionalAttendees);
+
+    Collection<TimeRange> possibleTimes = new ArrayList<>();
+    if (attendees.isEmpty()) {
+      // If there is no mandatory attendee, only consider availabilities of optional attendees.
+
+      // {@code eventTimes} is the list of current events' time ranges that need to be avoided when scheduling.
+      List<TimeRange> eventTimes = new ArrayList<>();
+      for (List<TimeRange> timeList : optionalAttendeesEventTimes.values()) {
+        eventTimes.addAll(timeList);
+      }
+      Collections.sort(eventTimes, TimeRange.ORDER_BY_START);
+      possibleTimes = findTimeRangeGaps(eventTimes, duration);
+    } else {
+      // If there are mandatory attendees, consider optional attendees who can possibly attend and ignore those who can't. 
+
+      List<TimeRange> eventTimes = new ArrayList<>(attendeesEventTimes);
+
+      // Find the time slot(s) that maximize the number of optional attendees who can attend.
+      // Count down from the maximum number of optional attendees and find all combinations of that number of attendees.
+      // Once we find a combination where scheduling is possible, we've found the maximal group of attendees.
+      for (int num = optionalAttendees.size(); num > 0; num--) {
+        String[] arr = optionalAttendees.toArray(new String[optionalAttendees.size()]);
```

Comment:
Can we rename this to something more relevant?

Reply:
okay